src/workflows/ingestion/make_reference/main.nf

```diff
@@ -1,142 +1,64 @@
-
-include { readConfig; channelFromParams; preprocessInputs; helpMessage } from workflowDir + "/utils/WorkflowHelper.nf"
-include { setWorkflowArguments; getWorkflowArguments } from workflowDir + "/utils/DataflowHelper.nf"
-
-config = readConfig("$workflowDir/ingestion/make_reference/config.vsh.yaml")
-
-workflow make_reference_entrypoint {
-  helpMessage(config)
-
-  channelFromParams(params, config)
-    | run_wf
-}
-
 workflow run_wf {
   take:
   input_ch
 
   main:
-  
-  ref_ch = input_ch
+
+  def targetMapping = [
+    "build_cellranger_reference": "cellranger",
+    "build_bdrhap_reference": "bd_rhapsody",
+    "star_build_reference": "star" 
+  ]
+
+  output_ch = input_ch
     // split params for downstream components
-    | preprocessInputs("config": config)
-    | setWorkflowArguments(
-      make_reference: [
+    | make_reference_component.run(
+      fromState: [
+        "input": "input",
         "genome_fasta": "genome_fasta", 
         "transcriptome_gtf": "transcriptome_gtf",
         "ercc": "ercc",
         "output_fasta": "output_fasta",
         "output_gtf": "output_gtf",
         "subset_regex": "subset_regex"
       ],
-      cellranger: [
-        "output": "output_cellranger",
-        "target": "target"
-      ],
-      bd_rhapsody: [
-        "output": "output_bd_rhapsody",
-        "target": "target"
+      toState: [
+        "output_fasta": "output_fasta",
+        "output_gtf": "output_gtf"
       ],
-      star: [
-        "output": "output_star",
-        "target": "target"
-      ]
+      auto: [publish: true]
     )
-
-    // generate reference
-    | getWorkflowArguments(key: "make_reference")
-    | make_reference.run(auto: [ publish: true ])
-
-
-  // generate cellranger index (if so desired)
-  cellranger_ch = ref_ch
-    | getWorkflowArguments(key: "cellranger")
-    | filter{ "cellranger" in it[1].target }
-    | build_cellranger_reference.run(
-      renameKeys: [ genome_fasta: "output_fasta", transcriptome_gtf: "output_gtf" ], 
-      auto: [ publish: true ]
-    )
-    | map{ tup -> tup.take(2) }
-
-  // generate bd_rhapsody index (if so desired)
-  bd_rhapsody = ref_ch
-    | getWorkflowArguments(key: "bd_rhapsody")
-    | filter{ "bd_rhapsody" in it[1].target }
-    | build_bdrhap_reference.run(
-      renameKeys: [ genome_fasta: "output_fasta", transcriptome_gtf: "output_gtf" ], 
-      auto: [ publish: true ]
-    )
-    | map{ tup -> tup.take(2) }
-
-  // generate star index (if so desired)
-  star = ref_ch
-    | getWorkflowArguments(key: "star")
-    | filter{ "star" in it[1].target }
-    | star_build_reference.run(
-      renameKeys: [ genome_fasta: "output_fasta", transcriptome_gtf: "output_gtf" ], 
-      auto: [ publish: true ]
+    | runEach(
+      components: [ 
+        build_cellranger_reference,
+        build_bdrhap_reference,
+        star_build_reference
+      ],
+      filter: { id, state, component ->
+        state.target.contains(targetMapping.get(component.name))
```

Comment:
third approach! Note that `component.name` will not be equal to the `config.functionality.name` when an alias is used.

Reply:
See 1a85c0bc8a98132957ff31e33d42ad3bde4b4314